PushSharp.Apple/AppleHttpPushChannel.cs

```diff
@@ -0,0 +1,210 @@
+                    message.Headers.Add("apns-topic", appleNotification.Type == AppleNotificationType.Voip ? $"{AppleHttpPushChannelSettings.BundleId}.voip" : AppleHttpPushChannelSettings.BundleId);
+                    //message.Headers.Add("apns-id", ); -> apple creates one if not provided
+                    //message.Headers.Add("apns-priority", ""); -> apple default 10, send immediately
+                    //message.Headers.Add("apns-collapse-id", ""); -> not used
+
+                    message.Content = new StringContent(appleNotification.Payload.ToString());
+                    response = _httpClient.SendAsync(message).ConfigureAwait(false).GetAwaiter().GetResult();
+                }
+
+                if (response.IsSuccessStatusCode)
+                {
+                    if (callback != null)
+                        callback(this, new SendNotificationResult(notification));
+                }
+                else
+                {
+                    var responseBody = response.Content.ReadAsStringAsync().ConfigureAwait(false).GetAwaiter().GetResult();
+                    Log.Error("Error during APNS Send with channel {0}: {1} -> Code {2} - {3}",
+                                        _channelInstanceId,
+                                        appleNotification.Identifier,
+                                        response.StatusCode,
+                                        string.IsNullOrEmpty(responseBody) ? "No response body" : responseBody);
+
+                    // Should we delete the device token when: HttpStatusCode.BadRequest - BadDeviceToken - The specified device token is invalid.Verify that the request contains a valid token and that the token matches the environment.
+                    // special case to call delete device token, means that device token it no longer available on apple side
+                    if (response.StatusCode == HttpStatusCode.Gone)
+                    {
+                        if (callback != null)
+                        {
+                            SendNotificationResult result;
+                            if (_appleSettings.EnableDeleteTokenOn410Response)
+                            {
+                                result = new SendNotificationResult(notification, false, new Exception("Device token no longer available on APNs."))
+                                {
+                                    IsSubscriptionExpired = true,
+                                    OldSubscriptionId = appleNotification.DeviceToken
+                                };
+                            }
+                            else
+                            {
+                                result = new SendNotificationResult(notification, false);
+                            }
+                            callback(this, result);
+                            return;
+                        }
+                    }
```

Comment:
this logic seems out of place in this class. Let whatever handles the callback decide what action to take.

Reply:
this is to cover the functionality of the feedback service, to be able to call the same callback handled by the feedback we need to specify those specific result properties

Reply:
Managing APNs tokens is outside the scope of this class, and outside the scope of this library as well. This class shouldn't make decisions about apns token storage. it should just fire the appropriate callback

Reply:
what i did instead is mapping the result once we get this specific status code, and now the delegate on PullApnsNotificationSender checks if we should end up deleting that token